### DISCUSSION TITLE
Updating options

### DISCUSSION THREAD

**Discussion by madodig** [TriliumNext/Trilium#4364] [Q&A] [answered]
Hi,

first of all, thanks for sharing this app! I've just recently found out about it and it's amazing!

I'm working on a widget that's changing app theme. I'm using following SQL query to alter the theme value:
`UPDATE options SET value = '${themeId}' WHERE name = 'theme';`
I've also tried this:
`UPDATE options SET value = '${themeId}', utcDateModified = '${utcDateModified}' WHERE name = 'theme';`

After the value is changed, the change is not reflected in the Options / Appearance (it's showing the theme that was active during app start). Trilium has to be closed and reopened (frontend reload is not enough) to reflect the change. I'm running Trilium desktop (v0.60.4 on Windows).

I've seen the following discussion regarding syncing: https://github.com/zadam/trilium/discussions/3586. OP there had issues with Web instance sync, but changes were reflected in Windows client. Regardless, I've tried "Force full sync" and "Fill entity changes records", but without success. I've also tried something like:
```
UPDATE entity_changes
SET hash = 'fake value N', utcDateChanged = '${utcDateModified}'
WHERE entityId  = '_optionsAppearance' AND entityName = 'note_contents';
```
_(I even tried setting changeId to something different if it makes sense)_

I couldn't find if the data is cached on app start or assigned to a variable that could be updated on-the-fly.  Does anyone know if this is possible (preferably without Frontend reload)? I don't need the reload because the widget's working fine and the UI is altered, it's just that the theme value under Options does not match the current theme.

Thanks!

**Comment by zadam** [comment, marked as answer]
Hi, updating the database via SQL doesn't notify the application, and the app doesn't constantly re-read everything from DB to see if something changed ...

The way to do this would be via backend script API.

Unfortunately there isn't an official API for working with options yet, but you can still access it via private API:

```javascript
const themeOption = api.__private.becca.getOption('theme');
themeOption.value = 'your_theme';
themeOption.save();
```

However, note that specifically theme change needs a frontend reload anyway, since it involves loading new CSS ...

**Reply by madodig** [reply]
Thank you, this is exactly what I needed.

I don't need to reload the frontend since the widget updates the CSS; it was the only option I had since I couldn't make the Frontend aware of the change. It turned out to be pretty fluid, so I'll keep that approach.

This is what it looks like so far, I will share it once I fix some minor issues and clean the code a bit... (_note that the clock next to the switch is just for reference, it is not a part of the widget_). You can toggle between light/dark modes of the current theme (if the mode exists), choose a theme from the dropdown, and (optionally) configure time range for the dark theme, so it is automatically changed based on the time of day. The switch itself is changing appearance based on the current theme.


Thanks again!


https://github.com/zadam/trilium/assets/14966219/e045c20b-b69f-4e0b-911a-269a7eb06294

**Reply by zadam** [reply]
Looks very nice!
